core-primitives/stf-state-handler/src/in_memory_state_file_io.rs

```diff
@@ -0,0 +1,384 @@
+/*
+	Copyright 2021 Integritee AG and Supercomputing Systems AG
+
+	Licensed under the Apache License, Version 2.0 (the "License");
+	you may not use this file except in compliance with the License.
+	You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+	Unless required by applicable law or agreed to in writing, software
+	distributed under the License is distributed on an "AS IS" BASIS,
+	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+	See the License for the specific language governing permissions and
+	limitations under the License.
+
+*/
+
+#[cfg(feature = "sgx")]
+use std::sync::SgxRwLock as RwLock;
+
+#[cfg(feature = "std")]
+use std::sync::RwLock;
+
+use crate::{
+	error::{Error, Result},
+	file_io::StateFileIo,
+};
+use codec::Encode;
+use itp_types::ShardIdentifier;
+use std::{
+	collections::HashMap,
+	hash::Hasher as HasherTrait,
+	string::{String, ToString},
+	vec::Vec,
+};
+
+type StateHash = u64;
+type ShardDirectory<State> = HashMap<String, (StateHash, State)>;
+type ShardsRootDirectory<State> = HashMap<ShardIdentifier, ShardDirectory<State>>;
+
+/// State file I/O using (unencrypted) in-memory representation of the state files.
+/// Uses u64 hash type. Can be used as mock for testing.
+#[derive(Default)]
+pub struct InMemoryStateFileIo<State, Hasher>
+where
+	State: Clone + Default + Encode,
+	Hasher: HasherTrait + Clone + Default,
+{
+	emulated_file_directory: RwLock<ShardsRootDirectory<State>>,
+	hasher: Hasher,
+}
+
+impl<State, Hasher> InMemoryStateFileIo<State, Hasher>
+where
+	State: Clone + Default + Encode,
+	Hasher: HasherTrait + Clone + Default,
+{
+	#[allow(unused)]
+	pub fn new(hash_function: Hasher, shards: &[ShardIdentifier]) -> Self {
+		let shard_hash_map: HashMap<_, _> =
+			shards.iter().map(|s| (*s, ShardDirectory::<State>::default())).collect();
+
+		InMemoryStateFileIo {
+			emulated_file_directory: RwLock::new(shard_hash_map),
+			hasher: hash_function,
+		}
+	}
+
+	#[cfg(test)]
+	pub fn get_files_for_shard(
+		&self,
+		shard_identifier: &ShardIdentifier,
+	) -> Result<HashMap<String, (StateHash, State)>> {
+		let files_lock = self.emulated_file_directory.read().map_err(|_| Error::LockPoisoning)?;
+		files_lock
+			.get(shard_identifier)
+			.cloned()
+			.ok_or_else(|| Error::InvalidShard(*shard_identifier))
+	}
+
+	fn compute_state_hash(&self, state: &State) -> StateHash {
```

Comment:
This is a mock (or stub) implementation, `State` is a generic type, with a `Hasher` that is provided. Because it's used for testing, the hash type usually is `u64`, so nothing complex that itself would have a hasher.